### DISCUSSION TITLE
咨询下1.1.1版本删除向量相关的问题

### DISCUSSION THREAD

**Discussion by SuoSiFire** [milvus-io/milvus#17618] [Q&A and General discussion] [answered]
目前我们的服务部署在对象存储的集群上，读写都通过http请求来实现。使用中遇到当删除量级比较大（768维向量，500qps总计删除500w左右）的时候，wal文件会堆积比较多，堆积10个左右。 当停止继续写入消息后，wal文件会比较快的清空。 想咨询下几个问题
1. 删除的流程是怎样的，看代码貌似是先会记录在delete_docs,并且修改bloom_filter,然后有个apply删除的过程进行真正删除？ 这个真正删除的过程是不是会比较慢，先下载文件，删除对应数据后再上传？ 这里可以描述下流程吗？ 
2. 删除wal文件貌似是在flush阶段进行的，flush又是定期执行的。 当我们停止写入消息后，wal堆积的文件会比较快的被清空。这是不是意味着wal文件已经被消费完，只是没有被删除。 那么如果flush是定期执行的，这里就不应该有堆积。这里没有太弄懂原因，求指点。 
3. 我这边搭了个单机节点验证，发现debug日志中有比较多的 MemTable::Serialize collection 768_test: Finished flushing (3.845014 seconds [3845.013668 ms]) ，看起来比较耗费时间，这个可能和wal文件堆积有关吗？

**Comment by SuoSiFire** [comment]
@yhmo  大神，有空帮忙看下

**Reply by SuoSiFire** [reply]
通过监控观察，开始删除操作后，milvus对于存储的读写qps都降低了，读写的rt都变高了。是不是删除过程中合并操作或者构建索引的操作耗时较长，导致其他的写操作有一定堆积

**Comment by SuoSiFire** [comment]
我们这边的场景是集合大小为5kw左右，每天会先删除500w左右的向量，之后再增加700w左右的向量。 所以猜测原因可能是这样

5kw的集合分为多个1G(index_file_size=1024)的segment,当删除请求过来时，会依次删除每个segment中的一些向量。删除完成后，插入请求过来，这些被删除的segment再次被填满，触发构建索引的逻辑，几乎所有的segment都会再次构建索引，构建索引导致wal消费堆积，wal文件数骤增，在segment都再次被填满后，wal正常消费，文件数恢复正常。 

大佬帮忙看下有没有什么建议？

**Comment by yhmo** [comment, marked as answer]
v1.1.1版本的删除流程：
首先我们知道，insert的数据在flush触发之前会保留在一个buffer里，flush之后这些数据就会产生segment并持久化。
那么当delete操作进来的时候，主要做两个事情：1.如果buffer里有找到这些向量，就把buffer的向量在内存里删去。 2.查看每个已持久化的segment，如果找到，则改写bloomfilter文件，并在delete_doc里把id append上去，并重写delete_doc文件；如果该segment已经加载在cache里，则把cache里那个delete_doc的数组也相应修改。
可以看到这第二个事情是很重的操作，如果太频繁是吃不消的。

在delete操作进来时，会先写入wal，wal文件的大小是256MB，写满一个就会新开一个。每个delete操作都带有操作类型，表名，分区名，delete id数组这些信息。每个操作以append的形式写入一个wal文件。

一个完整的delete业务逻辑：
1. 客户端调用delete，指定一个向量的id列表，要删除这些向量
2. 服务端接收到请求，做一些基本检查
3. wal把delete操作的信息以append的方式写入wal文件
4. 核心层依次处理wal所记录的操作，拿到delete的操作信息
5. 在insert buffer中查看是否有该表的数据，把delete id列表比对一下，如果找到，就在buffer里把对应的向量删除
6. delete调用返回给客户端
7. 等待后台flush操作或用户手动调用flush操作。当flush发生时，先拿到该表的所有segment信息，一个个遍历查看。对于每个segment，加载其bloomfilter文件，如果bollomfilter说该id不在这里，就继续查看下一个segment。如果bloomfilter说该id在这里，也不能全信它，加载该segment的id列表文件，一个个id比对确认是否真的在里面。在这个阶段，bloomfilter数据和id列表数据都会被放进cache。
8. 如果确认id在某个segment里，则修改cache中的bloomfilter数据，并重写bloomfiler文件，改写cache中的id列表，重写id列表文件。
10. 当一个wal文件中所有的操作都被执行完后，后台线程会稍后删除这个wal文件


insert的业务逻辑：
1. 客户端调用delete，输入一些向量
2. 服务端接受到请求，做一些基本检查
3. wal把这些向量记录成一个操作，并以append的方式写入wal文件
4. 核心层依次处理wal所记录的操作，核心层拿到insert操作信息
5. 把插入的向量放进insert buffer里，如果用户没有给id，则给每条向量产生一个id
6. insert调用返回客户端，把自动产生的id也返回
7. 等待后台flush操作或用户手动flush操作，把insert buffer里的数据持久化成segment

**Comment by yhmo** [comment]
可以看到，较重的业务发生在flush阶段，如果每次删一条向量就flush一次，系统会不堪重负。server_config.yaml里的auto_flush_interval默认是1秒，假设你每秒调用delete一次，那么milvus会每秒都要把所有的segment遍历一次。

比较好的做法是尽量把insert操作集中，比如把一小段时间里要删除的id凑够一批之后集中调用一次delete。

另外，每次delete一个id和每次delete多个id，后者的wal文件也会节省很多流量。

**Reply by SuoSiFire** [reply]
感谢回复。 听起来主要原因是flush造成的。我们没有手动调用flush操作，配置的auto_flush_interval为120s。也就意味着每120s需要把所有的segment遍历一遍用来进行删除操作。目前我们已经做了攒批操作，只是要删除的量级为500w左右，目前批处理大小大概为1k。即使这样貌似也会有比较长时间的wal文件堆积。  但看上去删除操作的qps是比较高的，是不是我这边应该把批处理大小再调大？ 让整体删除的次数变少会更好一些。

**Reply by yhmo** [reply]
120秒里大概会insert多少条，delete多少条？

随着一天天这种insert/delete的混合操作，会出现一种情况：每个segment里的delete_doc越来越多，segment的数量也越来越多，但向量总量变化其实并不大。insert新产生的segment里的delete_doc相对来说少一些，那些旧的segment里的delete_doc有可能会很多。这种情况一方面占用硬盘和内存变大了，另一方面导致搜索时性能下降(因为每次搜索时都要把过滤掉很多已被deleted的向量)，还有就是导致后续的delete操作重写delete_doc时越来越慢(因为delete_doc变大了)。

这需要在空闲的时候做compact操作，接口就叫compact(collection_name)，这个操作会检查所有的segment，如果某个segment里deleted的数量达到向量条数的10%以上，就会对其进行重建，建立一个新的segment以替换旧的，新的segment里会把deleted的向量都剔除掉。这是个很重的操作，最好在夜黑风高的时候做。这个操作也会重建索引，所以耗时很长。

**Reply by SuoSiFire** [reply]
120秒内大概会insert 12w条，delete6w条。 但是insert和delete不是并行的，会等到delete任务结束后再insert。

**Reply by yhmo** [reply]
看了一下compact的代码，把旧的segment瘦身成新的segment时，segment变小了，但数量变多了，因为compact操作不调用merge操作，而是直接为每个新的segment建索引。这样虽然问题也不大，不过segment数量多了还是会对search性能有些影响。如果非常在意search性能的，可以这样操作：
1. 调用drop_index把索引删除，这时只剩下原始向量数据
2. 调用compact，对所有segment做一次瘦身，产生了一堆新的segment
3. 调用create_index，重建索引，create_index会首先把小的segment合并，合并成较大的segment(理论上最大能接近到index_file_size的2倍)

这三步走下来，整个数据集就变得很清爽了，又可以欢快地用一段时间。过了若干天，发现又逐渐变慢了，就再次走这三步。

**Reply by SuoSiFire** [reply]
好的，目前检索性能看起来还可以接受。 只是wal的堆积问题对于索引的实时性影响比较大。我先按照上面说的尝试下调大删除操作的batch size，看下是不是会有所改善。

**Comment by SuoSiFire** [comment]
auto_flush_interval @yhmo  麻烦咨询下，如果这个字段设置为2min，是会导致插入的消息2min之后才能查询到吗？

**Reply by yhmo** [reply]
> auto_flush_interval @yhmo 麻烦咨询下，如果这个字段设置为2min，是会导致插入的消息2min之后才能查询到吗？

插入的数据先放在insert buffer里，在没有被flush落盘前都无法查询到。
flush触发的条件：
1. 后台自动flush，以auto_flush_interval为间隔
2. 客户端调用flush
3. insert buffer满了也会触发，insert buffer的大小在server_config.yaml 里设置，默认1GB
cache:
  insert_buffer_size: 1GB

4.调用create_index() 也会触发一次flush

**Reply by SuoSiFire** [reply]
了解了，感谢
